Add a parameter to cap the number of searches the msearch api will concurrently execute

martijnvg opened the pull request:
Before the msearch api would concurrently execute all searches concurrently. If many large
msearch requests would be executed this could lead to some searches being rejected
while other searches in the msearch request would succeed.

The goal of this `max_parallel_searches` parameter is to avoid this exhausting of the search TP.

martijnvg commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java:
```diff
@@ -38,6 +38,7 @@
  */
 public class MultiSearchRequest extends ActionRequest<MultiSearchRequest> implements CompositeIndicesRequest {
 
+    private int maxParallelSearchRequests = 10;
```
Is this a good default? I was also thinking about taking into account the number of active copies. (by picking the the index with lowest number of copies and multiply that by a constant factor)

clintongormley replied:
Perhaps the limit should be based on nodes so eg max of N concurrent search requests per node?

martijnvg replied:
+1 makes sense as a default

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -32,54 +33,102 @@
-                private void finishHim() {
-                    listener.onResponse(new MultiSearchResponse(responses.toArray(new MultiSearchResponse.Item[responses.length()])));
+                private void nextBatch() {
+                    if (counter.decrementAndGet() == 0) {
```
I think that we can do better than this. If I'm reading this correctly, this means that we wait until an entire batch is complete before submitting another batch of requests. Thus, a slow request can hold the next batch and thus the response. I think instead we should try to maintain as many requests in the queue as possible, up to the concurrent request limit.

martijnvg replied:
true, good point. I'll change the logic so that we don't wait until all requests have completed for the current batch and as long as there is a slot available the next search request will be executed.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java:
```diff
@@ -60,6 +61,25 @@ public MultiSearchRequest add(SearchRequest request) {
+    /**
+     * Returns the amount of search requests specified in this multi search requests are allowed to be ran in parallel.
+     */
+    public int maxParallelSearchRequests() {
```
Nit: I think it should the name should reflect concurrency rather than parallelism.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/action/search/RestMultiSearchAction.java:
```diff
@@ -91,6 +91,9 @@ public RestMultiSearchAction(Settings settings, RestController controller, Clien
     @Override
     public void handleRequest(final RestRequest request, final RestChannel channel, final Client client) throws Exception {
         MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        if (request.hasParam("max_parallel_searches")) {
```
Nit: I think that name of the request parameter should reflect concurrency rather than parallelism.

martijnvg commented:
@clintongormley @jasontedor I've updated this PR and changed how the msearch api executes concurrently and how the default number of search requests to execute concurrently is determined

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java:
```diff
@@ -60,6 +61,25 @@ public MultiSearchRequest add(SearchRequest request) {
+    /**
+     * Sets how many search requests specified in this multi search requests are allowed to be ran concurrently.
+     */
+    public MultiSearchRequest maxConcurrentSearchRequests(int maxParallelSearchRequests) {
```
Nit: `maxParallelSearchRequest` -> `maxConcurrentSearchRequests`

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java:
```diff
@@ -60,6 +61,25 @@ public MultiSearchRequest add(SearchRequest request) {
+     */
+    public MultiSearchRequest maxConcurrentSearchRequests(int maxParallelSearchRequests) {
+        if (maxParallelSearchRequests < 1) {
+            throw new IllegalArgumentException("maxConcurrentSearchRequests must be 1 or higher");
```
`maxConcurrentSearchRequests must be positive`?

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/MultiSearchRequestBuilder.java:
```diff
@@ -71,4 +71,12 @@ public MultiSearchRequestBuilder setIndicesOptions(IndicesOptions indicesOptions
+    /**
+     * Sets how many search requests specified in this multi search requests are allowed to be ran concurrently.
+     */
+    public MultiSearchRequestBuilder setMaxConcurrentSearchRequests(int maxParallelSearchRequests) {
```
Nit: `maxParallelSearchRequests` -> `maxConcurrentSearchRequests`

jasontedor commented on rest-api-spec/src/main/resources/rest-api-spec/api/msearch.json:
```diff
@@ -20,6 +20,10 @@
+        },
+        "max_concurrent_searches" : {
+          "type" : "number",
+          "description" : "Controls the maximum number of parallel searches the multi search api will execute"
```
Nit: `parallel` -> `concurrent`

jasontedor commented on docs/reference/search/multi-search.asciidoc:
```diff
@@ -71,6 +71,10 @@ against the `test2` index.
 all search requests.
 
+The msearch's `max_concurrent_searches` request parameter can be used to control
+the maximum number of parallel searches the multi search api will execute.
```
Nit: `parallel` -> `concurrent`

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -0,0 +1,138 @@
+                (Settings.EMPTY, "action", threadPool, actionFilters, resolver, taskManager) {
+            @Override
+            protected void doExecute(SearchRequest request, ActionListener<SearchResponse> listener) {
+                int currentParallelSearches = counter.incrementAndGet();
```
Nit: `currentParallelSearches` -> `currentConcurrentSearches`

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -0,0 +1,138 @@
+            protected void doExecute(SearchRequest request, ActionListener<SearchResponse> listener) {
+                int currentParallelSearches = counter.incrementAndGet();
+                if (currentParallelSearches > maxAllowedConcurrentSearches) {
+                    errorHolder.set(new AssertionError("Current parallel search [" + currentParallelSearches +
```
Nit: `parallel` -> `concurrent`

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -0,0 +1,138 @@
+            assertThat(response.getResponses().length, equalTo(numSearchRequests));
+            assertThat(errorHolder.get(), nullValue());
+        } finally {
+            threadPool.shutdownNow();
```
Just call `ESTestCase#terminate`?

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -22,64 +22,125 @@
+
+    /*
+     * This is not perfect and makes a big assumption, that all nodes have the same thread pool size / have the number
+     * of processors and that shard of the indices the search requests go to are more or less evenly distributed accross
```
Nit: `accross` -> `across`

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -22,64 +22,125 @@
+        });
+    }
+
+    final static class Item {
```
Can we give this a less generic name? `SearchRequestSlot`? `IndexedSearchRequest`?

martijnvg commented:
Thanks @jasontedor, I've updated the PR with this commit: e06b3107aaa8425450a2e4861720a72b915b1a4d

ebuildy commented:
What about a great multi search "pipeline" with parrallel and sequential execution? (I am always executing "fallback queries" if main query didn't return any result and search a clean way to achieve this..)

martijnvg commented:
@ebuildy That isn't what the msearch api is designed for. I guess the best place for the fallback logic is on the application side as this is very use case specific. (maybe if not enough hits are returned, or specific results don't show up another search request should be executed)

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/MultiSearchRequestTests.java:
```diff
@@ -167,6 +164,13 @@ public void testResponseErrorToXContent() throws IOException {
                 builder.string());
     }
 
+    public void testMaxParallelSearchRequests() {
```
One more nit, if we could rename this to `testMaxConcurrentSearchRequests`; I promise no additional nits on this PR. :smile:

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -22,64 +22,125 @@
+     * all nodes in the cluster. But I think it is a good enough default for most cases, if not then the default should be
+     * overwritten in the request itself.
+     */
+    static int defaultMaxConcurrentSearches(int availableProcessors, ClusterState state) {
```
Yeah, this is really tricky, but I think you did the right thing here, I don't see a better way.

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -22,64 +22,125 @@
+        int numDateNodes = state.getNodes().getDataNodes().size();
+        // availableProcessors will never be larger than 32, so max defaultMaxConcurrentSearches will never be larger than 49,
+        // but we don't know about about other search requests that are being executed so lets cap at 10 per node
+        int defaultSearchThreadPoolSize = Math.min(((availableProcessors * 3) / 2) + 1, 10);
```
Can we expose the default search thread pool size calculation as a static method in `ThreadPool.java` and defer to that here?

jasontedor commented on core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java:
```diff
@@ -22,64 +22,125 @@
-                        finishHim();
-                    }
-                }
+        Queue<SearchRequestSlot> searchRequestSlots = ConcurrentCollections.newQueue();
```
I'd rather just `new ConcurrentLinkedQueue<>()`.

jasontedor commented on core/src/test/java/org/elasticsearch/action/search/TransportMultiSearchActionTests.java:
```diff
@@ -0,0 +1,137 @@
+        IndexNameExpressionResolver resolver = new IndexNameExpressionResolver(Settings.EMPTY);
+
+        // Keep track of the number of concurrent searches started by multi search api,
+        // and if there are more searches than is afllowed create an error and remember that.
```
Typo: `afllowed` -> `allowed`

jasontedor commented:
I love it @martijnvg, I left a few minor comments and a small suggestion. No need for another round, fire at will.

martijnvg commented:
@jasontedor Thx for reviewing!

messense commented:
I have a dumb question. Is multi search executing concurrently on Elasticsearch 1.x?

martijnvg commented:
@messense Yes, however the number of searches executing concurrently wasn't capped, which could cause searches inside msearch request to fail (if the msearch contained many search requests), because some nodes couldn't handle the amount of searches being executed concurrently.
